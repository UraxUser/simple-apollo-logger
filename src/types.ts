import { Request } from 'apollo-server-env'
import { GraphQLRequestContext } from 'apollo-server-types'
import { DocumentNode } from 'graphql'

type TContext = any // workaround since types are in different file than extension class

export interface ApolloRequest {
  request: Pick<Request, 'url' | 'method' | 'headers'>
  queryString?: string
  parsedQuery?: DocumentNode
  operationName?: string
  variables?: { [key: string]: any }
  persistedQueryHit?: boolean
  persistedQueryRegister?: boolean
  context: TContext
  requestContext: GraphQLRequestContext<TContext>
}
