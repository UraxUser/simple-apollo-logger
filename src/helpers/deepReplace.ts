const deepReplace = (
  obj: any,
  searchedKey: string,
  replacementString: string
) => {
  if (typeof obj === 'object') {
    for (const [key, value] of Object.entries(obj)) {
      if (typeof value === 'object') {
        deepReplace(obj[key], searchedKey, replacementString)
      } else if (typeof value === 'string' && key === searchedKey) {
        obj[key] = replacementString
      }
    }
  }
}

export default deepReplace
