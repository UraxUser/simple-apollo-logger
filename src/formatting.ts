import deepReplace from './helpers/deepReplace'
import { ApolloRequest } from './types'

const filterPasswordFromVariables = (
  variables: any,
  keyword: string,
  replacementText: string
) => {
  // todo: make this function pure
  deepReplace(variables, keyword, replacementText)

  return variables
}

export const stringifiedRequestAttributes = (
  { variables, queryString, operationName }: ApolloRequest,
  variableFilter: VariableFilter | undefined | false
): string => {
  const copyOfVariables = JSON.parse(JSON.stringify(variables))
  if (variableFilter) {
    variableFilter.keywords.forEach(keyword => {
      filterPasswordFromVariables(
        copyOfVariables,
        keyword,
        variableFilter.replacementText
      )
    })
  }

  const stringifiedVariables = JSON.stringify(copyOfVariables)
  const stringifiedQueryString = JSON.stringify(queryString)
    .replace(/\s/g, '')
    .replace(/\\n/g, ' ')

  return `Request started
  Operation name: ${operationName}
  QueryString: ${stringifiedQueryString}
  Variables: ${stringifiedVariables}`
}

export interface VariableFilter {
  keywords: string[]
  replacementText: string
}
