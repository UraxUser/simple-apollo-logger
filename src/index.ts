import { stringifiedRequestAttributes } from './formatting'

export default class ApolloLogExtension {
    private options: Options
    private logger: Logger

    constructor(options: Options) {
        this.options = options
        this.logger = options.logger
    }

    requestDidStart(request: any): void {
        this.logger.log(stringifiedRequestAttributes(request))
    }
  
    willSendResponse(object: any): void {
      this.logger.log(JSON.stringify(object.graphqlResponse.data))
    }

    parsingDidStart(r: any): void {}
    validationDidStart(): void {}
    executionDidStart(r: any): void {}
}

export interface Options {
    logger: Logger
    logRequests: boolean
    logResponses: boolean
}

export interface Logger {
    log(msg: string): any
}