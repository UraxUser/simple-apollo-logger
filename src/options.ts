import { VariableFilter } from './formatting'

export const defaultOptions = {
  logger: (msg: string) => console.log(msg),
  logRequests: true,
  logResponses: false,
  prefix: () => `[${Date.now()}] `,
  variableFilter: {
    keywords: ['password'],
    replacementText: '[FILTERED]'
  },
  ignoreSchemaRequest: false
}

export interface Options {
  logger: (msg: string) => any
  logRequests: boolean
  logResponses: boolean
  prefix: () => string
  variableFilter: VariableFilter | false
  ignoreSchemaRequest: boolean
}

export interface UserOptions {
  logger?: (msg: string) => any
  logRequests?: boolean
  logResponses?: boolean
  prefix?: () => string
  variableFilter?: VariableFilter | false
  ignoreSchemaRequest?: boolean
}
