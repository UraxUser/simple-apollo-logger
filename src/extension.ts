import { stringifiedRequestAttributes } from './formatting'
import { defaultOptions, Options, UserOptions } from './options'
export default class ApolloLogExtension {

  private options: Options

  constructor(ops: UserOptions = {}) {
    this.options = Object.assign(defaultOptions, ops)
  }

  public requestDidStart(request: any): void {
    const isInspectionQuery = request.operationName === "IntrospectionQuery"
    const isIgnoredRequest = this.options.ignoreSchemaRequest && isInspectionQuery
    const shouldBeLogged = this.options.logRequests && !isIgnoredRequest
    if (shouldBeLogged) {
      this.log(
        stringifiedRequestAttributes(request, this.options.variableFilter)
      )
    }
  }

  public willSendResponse(object: any): void {
    if (this.options.logResponses) {
      this.log(JSON.stringify(object.graphqlResponse.data))
    }
  }

  // tslint:disable-next-line: no-empty
  public parsingDidStart(r: any): void {}
  // tslint:disable-next-line: no-empty
  public validationDidStart(): void {}
  // tslint:disable-next-line: no-empty
  public executionDidStart(r: any): void {}

  private log(msg: string): void {
    this.options.logger.log(`${this.options.prefix()} ${msg}`)
  }
}