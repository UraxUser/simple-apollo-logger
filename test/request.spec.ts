import { getObjectFromFile } from './helpers/jsonHelper'
import { request } from '../src/formatting'
import Logger from '../src/index'

const introspectionRequest = getObjectFromFile('data/requests/introspection.json') as request
const loginRequest = getObjectFromFile('data/requests/login.json') as request

const logger = new Logger()

console.log = jest.fn()
jest
    .spyOn(global.Date, "now")
    .mockImplementation(() => 
        new Date('2020-09-21T12:02:59.135Z').valueOf()
    )

describe("requestDidStart function", () => {
    describe('when logging requests is enabled', () =>{
        describe('introspection query request', () => {
            it('logs introspection query', () => {
                logger.requestDidStart(introspectionRequest)

                expect(console.log).toMatchSnapshot()
            })
        })

        describe('login mutation request', () => {
            it('logs login mutation', () => {
                logger.requestDidStart(loginRequest)

                expect(console.log).toMatchSnapshot()
            })
        })
    })
})
