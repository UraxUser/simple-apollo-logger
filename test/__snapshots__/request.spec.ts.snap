// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestDidStart function when logging requests is enabled introspection query request logs introspection query 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1600689779135] IntrospectionQuery
Headers: {\\"Authorization\\":\\"eyyyyLMAO.dd.json\\"}
QueryString: \\"queryIntrospectionQuery{ __schema{ queryType{ name } mutationType{ name } subscriptionType{ name } types{ ...FullType } directives{ name description locations args{ ...InputValue } } } }  fragmentFullTypeon__Type{ kind name description fields(includeDeprecated:true){ name description args{ ...InputValue } type{ ...TypeRef } isDeprecated deprecationReason } inputFields{ ...InputValue } interfaces{ ...TypeRef } enumValues(includeDeprecated:true){ name description isDeprecated deprecationReason } possibleTypes{ ...TypeRef } }  fragmentInputValueon__InputValue{ name description type{ ...TypeRef } defaultValue }  fragmentTypeRefon__Type{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name } } } } } } } } \\"
Variables: {}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`requestDidStart function when logging requests is enabled login mutation request logs login mutation 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1600689779135] IntrospectionQuery
Headers: {\\"Authorization\\":\\"eyyyyLMAO.dd.json\\"}
QueryString: \\"queryIntrospectionQuery{ __schema{ queryType{ name } mutationType{ name } subscriptionType{ name } types{ ...FullType } directives{ name description locations args{ ...InputValue } } } }  fragmentFullTypeon__Type{ kind name description fields(includeDeprecated:true){ name description args{ ...InputValue } type{ ...TypeRef } isDeprecated deprecationReason } inputFields{ ...InputValue } interfaces{ ...TypeRef } enumValues(includeDeprecated:true){ name description isDeprecated deprecationReason } possibleTypes{ ...TypeRef } }  fragmentInputValueon__InputValue{ name description type{ ...TypeRef } defaultValue }  fragmentTypeRefon__Type{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name ofType{ kind name } } } } } } } } \\"
Variables: {}",
    ],
    Array [
      "[1600689779135] null
Headers: undefined
QueryString: \\"mutation($input:SignupInput!){ signup(input:$input){ token } } \\"
Variables: {\\"input\\":{\\"email\\":\\"ddddddddassaaaaad@dddd.dde\\",\\"password\\":\\"[FILTERED]\\",\\"name\\":\\"ddoopson\\"}}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
